# idinf
Tools for disambiguating metagenomic reads

**phylowalk_m8parser.py**

Implements taxonomic tree reversal for .m8 files.

`python3 phylowalk_m8parser.py -i [input .m8 file] -o [output file]`


**align_and_convert_reference_microbe.sh**

For any input .fasta file, runs GSNAPL allowing 1000 matches and subsequently launches the phylowalk_m8parser.py script

`bash align_and_conver_reference_microbe.sh [fasta file name, not including .fasta]`

ie: `bash align_and_convert_reference_microbe.sh UGANDA/UGANDA_S9_NP_orthopneumovirus-hits`


**phylowak_SAMparser.py**

Implements taxonomic tree reversal for .sam files

`python3 phylowalk_SAMparser.py -i [input .sam file] -o [output file] [-ncbi] [-t]`

New options include:

-ncbi: If .sam file was generated by aligning to the full NCBI database, then set this flag to parse out GI from NCBI format "gi|XXX|gb|XXX|". Else, leave flag off to extract references as-is.

-t: If script is run on a server containing the appropriate NCBI database (hard-coded right now) for taxonomic reversal, then set this flag to do phylo-walking. Else, leave flag off to only analyze the raw GI content of multi-mapping reads. 

